File: GAPython/Chromosome.py
import random

class Chromosome:
    def __init__(self, **kwargs):
        self.length = kwargs["length"]
        self.fitness_fun = kwargs["fitness_fun"]
        if "genes" in kwargs:
            self.genes = kwargs["genes"]
        else:
            self.genes = random.sample([i for i in range(0, self.length)], self.length)
        self.fitness = self.fitness_fun(self.genes)

    def __str__(self):
        return str(self.genes)

    def crossover(self, other):
        parents = {"parent_1": self, "parent_2": other}
        new_genes_1 = [-1] * self.length
        new_genes_2 = [-1] * self.length

        # Define who has better fitness & assign better fitness to parent 1
        if self.fitness < other.fitness:
            parents["parent_1"], parents["parent_2"] = parents["parent_2"], parents["parent_1"]

        # Generate random start and random length of gene copied
        mut_length = random.randint(int(self.length * 0.05), self.length - 1)
        mut_start = random.randint(0, self.length - mut_length - 1)

        # Put part of parent one gene into new gene
        new_genes_1 = new_genes_1[:mut_start] + parents["parent_1"].genes[mut_start: mut_start + mut_length] + new_genes_1[mut_start + mut_length:]
        new_genes_2 = new_genes_2[:mut_start] + parents["parent_2"].genes[mut_start: mut_start + mut_length] + new_genes_2[mut_start + mut_length:]

        for i in parents["parent_2"].genes:
            if i not in new_genes_1:
                new_genes_1[new_genes_1.index(-1)] = i

        for j in parents["parent_1"].genes:
            if j not in new_genes_2:
                new_genes_2[new_genes_2.index(-1)] = j

        length = self.length
        fitness_fun = self.fitness_fun

        child_1 = Chromosome(length=length, genes=new_genes_1, fitness_fun=fitness_fun)
        child_2 = Chromosome(length=length, genes=new_genes_2, fitness_fun=fitness_fun)
        return (child_1, child_2)

    def mutate(self):
        swap_1 = random.randint(0, self.length - 1)
        swap_2 = random.randint(0, self.length - 1)
        length = random.randint(int((self.length - swap_2) * 0.02), self.length - swap_2)

File: Chromosome/Chromosome.cpp
#include "Chromosome.h"
#include <iostream>
#include <string>
#include <vector>
#include <stdio.h>
#include <ctype.h>
#include <cctype>
#include <map>
#include <functional>
#include <numeric>
#include <random>
#include <iterator>
#include <algorithm>
#include <time.h>
#include <functional>
#include <sstream>
#include "functions.h"
#include "difflib.h"


using namespace std;

void Chromosome::Init(int len, std::vector<int> gene_vect) {
    this->length = len;
//    fitness_func = fit_fumc;
    if(gene_vect.size()){
        this->genes = gene_vect;
    }else{
        vector<int> x;

        for(int i = 0; i<(*this).length; i++){
            x.push_back(rand()%(*this).length);
        }

        this->genes = x;
    }
//    this->fitness = fitness_func((*this).genes);
}

vector<Chromosome> Chromosome::crossover(Chromosome *other_chrom) {
    map<string,Chromosome*> parents = {{"parent1",this}, {"parent2", other_chrom}};
//    cout<<(*this).length<<endl;
    vector<int> new_genes_1;
    vector<int> new_genes_2;

    int part_1 = rand() % ((*this).genes.size()) + 0;

    // Put part of parent one gene into new gene
    vector<vector<int>> container1 = {slicer((*parents["parent1"]).genes.begin(), (*parents["parent1"]).genes.begin() + part_1 + 1), slicer((*parents["parent2"]).genes.begin() + part_1, (*parents["parent2"]).genes.begin() + (*parents["parent1"]).genes.size())};
    v_merge(new_genes_1, container1);

    vector<vector<int>> container2 = {slicer((*parents["parent2"]).genes.begin(), (*parents["parent2"]).genes.begin() + part_1 + 1), slicer((*parents["parent1"]).genes.begin() + part_1, (*parents["parent1"]).genes.begin() + (*parents["parent2"]).genes.size())};
    v_merge(new_genes_2, container2);

    int new_length = (*this).length;
    function<int(vector<int>)> new_fitness_func = (*this).fitness_func;

    Chromosome child_1 = Chromosome(new_length, new_fitness_func, new_genes_1);
    Chromosome child_2 = Chromosome(new_length, new_fitness_func, new_genes_2);

    return vector<Chromosome>{child_1,child_2};
}
void Chromosome::mutate() {
    int swap_1 = rand() % ((*this).genes.size()) + 0;
    int swap_2 = rand() % ((*this).genes.size()) + 0;

    swap(this->genes[swap_1], this->genes[swap_2]);
}

void Chromosome::calc_fitnes() {
//    cout << "here2"<< endl;

    fitness = fitness_func(genes);

//    cout << "here3"<< endl;
}


void Chromosome::calc_ratio(Chromosome best_chromosome) {
    string a1 = vector_to_string((*this).genes);
    string a2 = vector_to_string(best_chromosome.genes);

    auto matcher = difflib::MakeSequenceMatcher(a1,a2);

    ratio = matcher.ratio();

}


std::string Chromosome::vector_to_string(std::vector<int> genes) {
    std::ostringstream oss;
    if (!genes.empty())
    {
        std::copy(genes.begin(), genes.end()-1,
                  std::ostream_iterator<int>(oss));

        oss << genes.back();
    }
    return oss.str();
}

Chromosome::Chromosome() {

}
